pkg/gauges/contracts/LiquidityGaugeFactory.sol

```diff
@@ -0,0 +1,139 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/Authentication.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Clones.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Ownable.sol";
+import "@balancer-labs/v2-vault/contracts/interfaces/IVault.sol";
+
+import "./interfaces/ILiquidityGauge.sol";
+
+contract LiquidityGaugeFactory is Authentication, Ownable {
+    IVault private immutable _vault;
+    ILiquidityGauge private _gaugeImplementation;
+
+    mapping(address => bool) private _isGaugeFromFactory;
+    mapping(address => address) private _poolGauge;
+
+    event GaugeCreated(address indexed gauge, address indexed pool);
+    event GaugeImplementationUpdated(address oldGaugeImplementation, address newGaugeImplementation);
+
+    constructor(
+        IVault vault,
+        ILiquidityGauge gauge,
+        address authorizerAdaptor
+    ) Authentication(bytes32(uint256(address(this)))) {
+        _vault = vault;
+        _gaugeImplementation = gauge;
+        _transferOwnership(authorizerAdaptor);
+    }
+
+    /**
+     * @dev Returns the address of the Vault.
+     */
+    function getVault() public view returns (IVault) {
+        return _vault;
+    }
+
+    /**
+     * @dev Returns the address of the Vault's Authorizer.
+     */
+    function getAuthorizer() public view returns (IAuthorizer) {
+        return getVault().getAuthorizer();
+    }
+
+    /**
+     * @notice Returns the address of the Authorizer adaptor contract.
+     * @dev This address has full admin powers over all gauges deployed by this factory.
+     */
+    function getAuthorizerAdaptor() external view returns (address) {
+        return owner();
+    }
+
+    /**
+     * @notice Returns the address of the implementation used for gauge deployments.
+     */
+    function getGaugeImplementation() public view returns (ILiquidityGauge) {
+        return _gaugeImplementation;
+    }
+
+    /**
+     * @notice Returns the address of the gauge belonging to `pool`.
+     */
+    function getPoolGauge(address pool) external view returns (ILiquidityGauge) {
+        return ILiquidityGauge(_poolGauge[pool]);
+    }
+
+    /**
+     * @notice Returns true if `gauge` was created by this factory.
+     */
+    function isGaugeFromFactory(address gauge) external view returns (bool) {
+        return _isGaugeFromFactory[gauge];
+    }
+
+    /**
+     * @notice Sets the address of the gauge implementation to be used for future deployments
+     */
+    function setGaugeImplementation(address newGaugeImplementation) external authenticate {
+        address currentGaugeImplementation = address(_gaugeImplementation);
+
+        _gaugeImplementation = ILiquidityGauge(newGaugeImplementation);
+        emit GaugeImplementationUpdated(currentGaugeImplementation, newGaugeImplementation);
+    }
```

Comment:
Couldn't we simply deploy a new factory if we needed to do this?

Reply:
We could however that opens the possibility of having duplicate gauges for the same pool get deployed. We could add a killswitch to this factory to prevent it deploying gauges for pools from after the migration, however the new factory could deploy new gauges for old pools.

The only way to prevent this would be to always call into the previous factories to check if a gauges already exists.

Reply:
Right, I see the coordination problem now.

Given that the likelihood of us having to deploy a new gauge contract, or of the gauges being somehow faulty are both very low, I think the current design makes sense. Thanks!